crates/iota-grpc-api/src/types.rs

```diff
@@ -235,24 +242,144 @@ impl Reader {
+                        // Channel lagged - historical data should catch up via storage
                     }
                     Err(tokio::sync::broadcast::error::RecvError::Closed) => {
-                        // report internal error to the stream and break
                         Err(Status::internal("Checkpoint data channel closed."))?;
                         break;
                     },
                 }
+
+                // Update latest checkpoint info
                 latest = reader.get_latest().unwrap_or(start);
-                tracing::debug!("[profile][grpc] Updating latest checkpoint index to {latest}.");
+                tracing::debug!("[profile][grpc] Updated latest checkpoint index to {latest}.");
             }
         }
     }
 }
 
+/// Wrapper around RestStateReader that implements the GrpcStateReader trait
+#[derive(Clone)]
+pub struct GrpcReader {
+    state_reader: Arc<dyn GrpcStateReader>,
+}
+
+impl GrpcReader {
+    pub fn new(state_reader: Arc<dyn GrpcStateReader>) -> Self {
+        Self { state_reader }
+    }
+
+    pub fn from_rest_state_reader(state_reader: Arc<dyn RestStateReader>) -> Self {
+        Self {
+            state_reader: Arc::new(RestStateReaderAdapter {
+                inner: state_reader,
+            }),
+        }
+    }
+
+    pub fn get_epoch_last_checkpoint(
+        &self,
+        epoch: u64,
+    ) -> Result<Option<CertifiedCheckpointSummary>, Box<dyn std::error::Error + Send + Sync>> {
+        self.state_reader.get_epoch_last_checkpoint(epoch)
+    }
+}
+
+/// Adapter that implements GrpcStateReader for RestStateReader
+pub struct RestStateReaderAdapter {
+    inner: Arc<dyn RestStateReader>,
+}
+
+impl GrpcStateReader for RestStateReaderAdapter {
+    fn get_latest_checkpoint_sequence(&self) -> Option<u64> {
+        Some(*self.inner.get_latest_checkpoint().sequence_number())
+    }
+
+    fn get_checkpoint_summary(&self, seq: u64) -> Option<CertifiedCheckpointSummary> {
+        self.inner
+            .get_checkpoint_by_sequence_number(seq)
+            .map(CertifiedCheckpointSummary::from)
+    }
+
+    fn get_checkpoint_data(&self, seq: u64) -> Option<CheckpointData> {
+        let summary = self.inner.get_checkpoint_by_sequence_number(seq)?;
+        let contents = self.inner.get_checkpoint_contents_by_sequence_number(seq)?;
+        Some(self.inner.get_checkpoint_data(summary, contents))
+    }
+
+    fn get_epoch_last_checkpoint(
+        &self,
+        epoch: u64,
+    ) -> Result<Option<CertifiedCheckpointSummary>, Box<dyn std::error::Error + Send + Sync>> {
+        match self.inner.get_epoch_last_checkpoint(epoch) {
+            Ok(Some(checkpoint)) => Ok(Some(CertifiedCheckpointSummary::from(checkpoint))),
+            Ok(None) => Ok(None),
+            Err(e) => Err(Box::new(e) as Box<dyn std::error::Error + Send + Sync>),
+        }
+    }
+}
+
+#[derive(Clone)]
+pub struct Reader {
+    pub state_reader: Arc<dyn RestStateReader>,
+}
```

Comment:
What is the difference between `Reader` and `GrpcReader`? I also didn't see where `Reader` is used.